cmd/snap-repair/runner.go

```diff
@@ -64,22 +76,139 @@ func (r *Repair) SetStatus(status RepairStatus) {
+	rundir := r.RunDir()
 	err := os.MkdirAll(rundir, 0775)
 	if err != nil {
 		return err
 	}
+
 	script := filepath.Join(rundir, fmt.Sprintf("script.r%d", r.Revision()))
-	err = osutil.AtomicWriteFile(script, r.Body(), 0600, 0)
+	err = osutil.AtomicWriteFile(script, r.Body(), 0700, 0)
+	if err != nil {
+		return err
+	}
+
+	// the date may be broken so we use an additional counter
+	counter := nextCounter(r)
+	now := time.Now().Format("2006-01-02T150405")
+	baseName := fmt.Sprintf("r%d.%06d.%v", r.Revision(), counter, now)
+
+	logPath := filepath.Join(rundir, baseName+".output")
+	logf, err := os.OpenFile(logPath, os.O_RDWR|os.O_CREATE, 0600)
+	if err != nil {
+		return err
+	}
+	defer logf.Close()
+
+	// run things and captures output etc
+	stR, stW, err := os.Pipe()
 	if err != nil {
 		return err
 	}
+	defer stR.Close()
+	defer stW.Close()
+
+	// run the script
+	env := os.Environ()
+	// we need to hardcode FD=3 because this is the FD after
+	// exec.Command() forked. there is no way in go currently
+	// to run something right after fork() in the child to
+	// know the fd. However because go will close all fds
+	// except the ones in "cmd.ExtraFiles" we are safe to set "3"
+	env = append(env, "SNAP_REPAIR_STATUS_FD=3")
+	env = append(env, "SNAP_REPAIR_RUN_DIR="+rundir)
+
+	workdir := filepath.Join(rundir, "work")
+	if err := os.MkdirAll(workdir, 0700); err != nil {
+		return err
+	}
+
+	cmd := exec.Command(script)
+	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+	cmd.Env = env
+	cmd.Dir = workdir
+	cmd.ExtraFiles = []*os.File{stW}
+	cmd.Stdout = logf
+	cmd.Stderr = logf
+	if err = cmd.Start(); err != nil {
+		return err
+	}
+	stW.Close()
+
+	// wait for repair to finish or timeout
+	killTimerCh := time.After(defaultRepairTimeout)
+	doneCh := make(chan struct{})
+	go func() {
+		// ignore error, we only care about what we got via
+		// the status pipe
+		cmd.Wait()
+		close(doneCh)
+	}()
+	select {
+	case <-doneCh:
+		// done
+	case <-killTimerCh:
+		if err := osutil.KillProcessGroup(cmd); err != nil {
+			logger.Noticef("cannot kill timed out repair %s: %s", r, err)
+		}
+		statusPath := filepath.Join(rundir, baseName+".retry")
+		if err := osutil.AtomicWriteFile(statusPath, nil, 0600, 0); err != nil {
+			logger.Noticef("cannot write stamp: %s", statusPath)
+		}
+		return fmt.Errorf("repair did not finish within %s", defaultRepairTimeout)
+	}
+
+	// read repair status pipe, use the last value
+	var status RepairStatus
+	scanner := bufio.NewScanner(stR)
+	for scanner.Scan() {
+		switch strings.TrimSpace(scanner.Text()) {
+		case "done":
+			status = DoneStatus
+		case "skip":
+			status = SkipStatus
+		}
+	}
+	if scanner.Err() != nil {
+		return err
+	}
 
-	// XXX actually run things and captures output etc
+	// write stamp and set status
+	statusPath := filepath.Join(rundir, baseName+"."+status.String())
+	if err := osutil.AtomicWriteFile(statusPath, nil, 0600, 0); err != nil {
```

Comment:
Should we atomically rename the output file instead, into its respective status? This way we wouldn't need two files.

That said, it's also not clear why we have that logic of writing the status into the file individually.